FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.JobManagement/AzureStorageQueueClient.cs

```diff
@@ -0,0 +1,682 @@
+            var message = await _azureJobMessageQueueClient.ReceiveMessageAsync(visibilityTimeOut, cancellationToken);
+
+            if (message.Value == null)
+            {
+                _logger.LogWarning("Failed to receive message from queue, the queue message is null.");
+                return null;
+            }
+
+            var jobMessage = JsonConvert.DeserializeObject<JobMessage>(message.Value.Body.ToString());
+
+            if (jobMessage == null)
+            {
+                _logger.LogWarning("Failed to deserialize message, the deserialized job message is null.");
+                return null;
+            }
+
+            // step 2: get jobInfo entity to check job status, delete this message if job is already completed/failed/cancelled
+            // if status is running and CancelRequest is true, the job will be dequeued, and jobHosting will continue to handle it
+            var jobInfoEntityResponse = await _azureJobInfoTableClient.GetEntityAsync<JobInfoEntity>(
+                jobMessage.PartitionKey,
+                jobMessage.RowKey,
+                cancellationToken: cancellationToken);
+            var jobInfo = jobInfoEntityResponse.Value.ToJobInfo<TJobInfo>();
+
+            if (jobInfo.Status is JobStatus.Completed or JobStatus.Failed or JobStatus.Cancelled)
+            {
+                _logger.LogWarning($"Discard queue message {message.Value.MessageId}, the job status is {jobInfo.Status}.");
+                await _azureJobMessageQueueClient.DeleteMessageAsync(message.Value.MessageId, message.Value.PopReceipt, cancellationToken);
+
+                // dequeue next job
+                return await DequeueAsync(queueType, worker, heartbeatTimeoutSec, cancellationToken);
+            }
+
+            // step 3: get job lock entity to check message id
+            var jobLockEntityResponse = await _azureJobInfoTableClient.GetEntityAsync<JobLockEntity>(
+                jobMessage.PartitionKey,
+                AzureStorageKeyProvider.JobLockRowKey(jobInfo.JobIdentifier()),
+                cancellationToken: cancellationToken);
+
+            var jobLockEntity = jobLockEntityResponse.Value;
+
+            if (!string.Equals(jobLockEntity.JobMessageId, message.Value.MessageId, StringComparison.OrdinalIgnoreCase))
+            {
+                _logger.LogWarning($"Discard queue message {message.Value.MessageId}, the message id is inconsistent with the one in the table entity.");
+                await _azureJobMessageQueueClient.DeleteMessageAsync(message.Value.MessageId, message.Value.PopReceipt, cancellationToken);
+
+                // dequeue next job
+                return await DequeueAsync(queueType, worker, heartbeatTimeoutSec, cancellationToken);
+            }
+
+            // step 4: skip it if the job is running and still active
+            if (jobInfo.Status == JobStatus.Running && jobInfo.HeartbeatDateTime.AddSeconds(heartbeatTimeoutSec) > DateTime.UtcNow)
```

Comment:
this is also should not happen?

Reply:
This case happens in the scenario: update jobInfo'a heartbeat successfully while failing to update message visibility timeout when `KeepAliveJob`, then the message will be de queue again.

Reply:
Per our discussion, will throw an exception for it.